### ISSUE TITLE
Improve partitioning of tests across test jobs (slow, parallel, disruptive etc)

### ISSUE THREAD

**Issue by quinton-hoole** [kubernetes/kubernetes#13062]
Some e2e tests are slow by necessity and design.  Others are slow through non-optimal implementation.  We don't want these to slow down the completion of the faster e2e tests.  

The former group (slow by necessity) need to go into their own group in perpetuity.  The latter need to go into quarantine until they can be speeded up.  As a first step the two groups can share one test job.

This is not just a theoretical problem.  Right now, e2e tests are often timing out after 2 hours.  Not long ago they completed in 30 minutes.

**Comment by quinton-hoole** [comment]
The slowest e2e tests at the moment seem to be:

* 8 min 3 sec SchedulerPredicates validates MaxPods limit number of pods that are allowed to run. (Failed 6 times in the last 30 runs. Stability: 80 %) No flake or failure
* 6 min 15 sec Nodes Network when a minion node becomes unreachable [replication controller] recreates pods scheduled on the unreachable minion node AND allows scheduling of pods on a minion after it rejoins the cluster (No known failures. Stability 100 %) No flake or failure		Passed
* 4 min 50 sec Services should be able to change the type and nodeport settings of a service (No known failures. Stability 100 %) No flake or failure
* 3 min 48 sec Nodes Resize should be able to add nodes (No known failures. Stability 100 %) No flake or failure	3 min 48 sec	Passed
* 3 min 30 sec Nodes Resize should be able to delete nodes (Failed 1 times in the last 5 runs. Stability: 80 %) No flake or failure		Passed
* 2 min 35 sec Services should release the load balancer when Type goes from LoadBalancer -> NodePort (No known failures. Stability 100 %) No flake or failure	
* 2 min 43 sec Pod Disks should schedule a pod w/ a RW PD shared between multiple containers, write to PD, delete pod, verify contents, and repeat in rapid succession (No known failures. Stability 100 %) No flake or failure		Passed
* 2 min 17 sec Pod Disks should schedule a pod w/ a RW PD, remove it, then schedule it on another host (No known failures. Stability 100 %) No flake or failure	Passed
* 2 min 10 sec Pod Disks should schedule a pod w/ a readonly PD on two hosts, then remove both. (No known failures. Stability 100 %) No flake or failure	Passed
* 2 min 25 sec	 Pods should *not* be restarted with a /healthz http liveness probe (No known failures. Stability 100 %) No flake or failure	Passed
* 2 min 25 sec	Pods should *not* be restarted with a docker exec "cat /tmp/health" liveness probe (No known failures. Stability 100 %) No flake or failure	Passed
* 1min 36 s Cluster level logging using Elasticsearch should check that logs from pods on all nodes are ingested into Elasticsearch
* 1 min 38 s Etcd failure should recover from SIGKILL (this should move to the destructive "reboot" test job)
* 1 min 29 sec Etcd failure should recover from network partition with master (as above)
* 1 min 57 sec Namespaces Delete 90 percent of 100 namespace in 150 seconds

**Comment by quinton-hoole** [comment]
@ixdy @zmerlynn FYI

**Comment by roberthbailey** [comment]
An alternate approach would be to make all e2es runnable in parallel. Then the full e2e suite would take the time of the longest test run, rather than the sum of all test runs.

**Comment by quinton-hoole** [comment]
In theory yes, in practise it creates noise to run other tests in parallel with e.g. rebooting all the nodes or etcd, inducing network partitions, or saturating all nodes with pods.  There are some tests that are both worth running, and not sensible to run concurrently with other tests.

**Comment by quinton-hoole** [comment]
Thinking a little harder, I guess it might make sense to run as many tests as feasible in parallel only, and disable them in the serial test runs, so that those run a lot faster.  Mmmm...

**Comment by roberthbailey** [comment]
It would actually be really good to see which tests fail when run on a cluster that is resizing, because if we could change our tests to be resilient to resizes, then we could verify that our cluster continues working properly before/during/after a resize.

**Comment by quinton-hoole** [comment]
Yes, agreed.  I think that should be a separate test job, specifically testing cluster resizing.  Otherwise a failure in the cluster resize logic could manifest in assorted unrelated errors in our more general e2e tests.

**Comment by wojtek-t** [comment]
I think a huge number of tests may fail on cluster downsizing. The reason is that in many tests we are e.g. starting a pod and wait until this particular pod is running (and possibly running after we do something). And obviously downsizing the cluster can kill a pod running on a node that is being removed.

I agree that we should take advantage of replication controllers, but I think it's not easy to test some particular features assuming that a pod can die an be recreated.


Also, I do NOT think that running fast and parallelizable test only in "parallel" mode is a good idea. Running all the tests sequentially in a single suite gives us value, because some bugs appear only in some strange corner cases.
To give you an example, when I was switching on "watch in apiserver", it was passing all the "parallel tests" in a stable way, but after merging it started failing. It appeared that there is a deadlock, but this deadlock was only triggered when etcd was restarted/unavailable for some time. So the same test would pass in a parallel mode, but wasn't passing if run after test that was killing etcd.
I'm pretty sure there were other such cases.

Also note, that we already have "gce-parallel" suite where we are running tests that we think are safe to run in parallel.

**Comment by zmerlynn** [comment]
Counterproposal: We start balancing tests into job by a string hash, and
split the jobs up into e2e-1, e2e-2, etc. For bonus points, the salt of the
hash could change at an arbitrary (daily) period so the tests didn't always
run together.

**Comment by wojtek-t** [comment]
@zmerlynn - I like this idea as long as this will be changing frequently (I think more than once a day).

**Comment by zmerlynn** [comment]
That's fine, but may slightly up the complexity. I was proposing daily
because each job can then just use the shell date command as the seed for
the hash, with just, say, the day. Using the hour might work similarly well
as long as the jobs are short enough.

(I'm imagining just a minor pull to gingko that would hash and select which
portion, based on command line flags.)

**Comment by quinton-hoole** [comment]
I think that we need two types of test jobs:

1. Reproducible jobs (that run the same tests every time, possibly in the same order, and do not include tests that have a high probability of causing difficult-to-diagnose failures in other tests, like cluster resizes, machine and network failure simulations etc).  These are the primary regression canaries.
2. "Chaos" jobs, which mix tests up in different combinations and sequences in an attempt to increase the total area of test coverage.  These are designed to uncover unusual failures resulting from particular sequences of events.

**Comment by zmerlynn** [comment]
What I just described fully meets #1 w.r.t. test order/selection. You'd print out the selection seed, ginkgo would still print out it's orderseed for test randomization, and it's all reproducible selection/ordering.

I'm not sure I necessarily agree, though, that if you're executing tests serially (but sharded across clusters, as proposed) that you can't execute the resize tests or things like that. I actually think that's one of the huge potential advantages for sharding is that we could potentially add *all* the tests (including the 1000 reboot tests) back in.

**Comment by quinton-hoole** [comment]
There is still good discussion in this issue, not related to the title ("Label and move slow tests..."), and which still needs to be actioned.  Re-opening, until these have all been closed out or moved to separate issues.

**Comment by quinton-hoole** [comment]
Removing flake label, as this issue is not tracking a flaky test.  In it's current for it's tracking proposed improvements to test infrastructure.  Will rename issue accordingly.

**Comment by quinton-hoole** [comment]
Removing assignment to me.  Leave to the new test team to prioritize and assign.

**Comment by quinton-hoole** [comment]
@kubernetes/goog-testing @ixdy @fejta

**Comment by ihmccreery** [comment]
This falls squarely in the test-job assignment work I've been thinking about, so I'll take it for now.  Just filed #17596 which is exactly in here, too.

**Comment by ihmccreery** [comment]
Here's a rough outline of what needs to be done:

- [ ] Source controlled jenkins job config
- [ ] Goldenfiles for test -> jobs ("what jobs run a given test?") and job -> tests ("what tests does a given job run?")
- [ ] Structured metadata for tests / suite
- [ ] Move to auto assignment of tests to jobs based on metadata

**Comment by ihmccreery** [comment]
See @deads2k's https://github.com/kubernetes/kubernetes/issues/18845#issuecomment-165780104, too, about tagging tests for test before merge.

**Comment by ihmccreery** [comment]
Closing in favor of #10548.
